src/sql/src/session/vars2.rs

```diff
@@ -0,0 +1,685 @@
+        self.validate_constraints(v.as_ref())?;
+        self.default_value = Some(v);
+        Ok(())
+    }
+
+    pub fn reset(&mut self, local: bool) {
+        let value = self
+            .default_value
+            .as_ref()
+            .map(|v| v.as_ref())
+            .unwrap_or(self.definition.value.value());
+        if local {
+            self.local_value = Some(value.box_clone());
+        } else {
+            self.local_value = None;
+            self.staged_value = Some(value.box_clone());
+        }
+    }
+
+    pub fn end_transaction(&mut self, action: EndTransactionAction) {
+        self.local_value = None;
+        match action {
+            EndTransactionAction::Commit if self.staged_value.is_some() => {
+                self.session_value = self.staged_value.take()
+            }
+            _ => self.staged_value = None,
+        }
+    }
+
+    pub fn value_dyn(&self) -> &dyn Value {
+        self.local_value
+            .as_deref()
+            .or_else(|| self.staged_value.as_deref())
+            .or_else(|| self.session_value.as_deref())
+            .or_else(|| self.default_value.as_deref())
+            .unwrap_or(self.definition.value.value())
+    }
+
+    pub fn value<V: 'static>(&self) -> &V {
+        let val = self.value_dyn();
+        val.as_any().downcast_ref::<V>().expect("success")
+    }
+
+    fn validate_constraints(&self, val: &dyn Value) -> Result<(), VarError> {
+        if let Some(constraint) = &self.definition.constraint {
+            constraint
+                .check_constraint(self.value_dyn(), val)
+                .map_err(|kind| VarError {
+                    name: self.definition.name.as_str(),
+                    value: val.format(),
+                    type_name: (self.definition.type_name)(),
+                    kind,
+                })
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Var for SessionVar {
+    fn name(&self) -> &'static str {
+        self.definition.name.as_str()
+    }
+
+    fn value(&self) -> String {
+        self.value_dyn().format()
+    }
+
+    fn description(&self) -> &'static str {
+        self.definition.description
+    }
+
+    fn type_name(&self) -> String {
+        (self.definition.type_name)().to_string()
+    }
+
+    fn visible(
+        &self,
+        user: &User,
+        system_vars: Option<&super::vars::SystemVars>,
+    ) -> Result<(), super::vars::VarError> {
+        self.definition.visible(user, system_vars)
+    }
+}
+
+/// Session variables.
+///
+/// See the [`crate::session::vars`] module documentation for more details on the
+/// Materialize configuration model.
+#[derive(Debug, Clone)]
+pub struct SessionVars {
+    /// The set of all session variables.
+    vars: BTreeMap<&'static UncasedStr, SessionVar>,
+    /// Inputs to computed variables.
+    build_info: &'static BuildInfo,
+    /// Information about the user associated with this Session.
+    user: User,
+}
```

Comment:
How do you plan on implementing the variable specific functions that give back concrete types. For example:

```
    /// Returns the value of the `idle_in_transaction_session_timeout` configuration parameter.
    pub fn idle_in_transaction_session_timeout(&self) -> &Duration {
       ...
    }
```

Reply:
Same as they exist today! We retain `fn expect_value(...)` where the caller needs to specify the type we downcast to.

It is a little more fragile than before though, because the compiler can't check that the type is correct. This is the only downside to the new approach. I'm not too concerned though because it's a surface area that is well tested and only needs to get written once. Also I tried using const generics to get the same compiler-checked safety and you can get close with the `adt_const_params` nightly feature, but I ran into compiler crashes